src/macros-by-example.md

```diff
@@ -24,149 +24,404 @@
+  * `literal`: matches `-`<sup>?</sup>[_LiteralExpression_]
+
+In the transcriber, metavariables are referred to simply by `$`_name_`, since
+the fragment kind is specified in the matcher. Metavariables are replaced with
+the syntax element that matched them. The keyword metavariable `$crate` can be
+used to refer to the current crate; see [Hygiene] below. Metavariables can be
+transcribed more than once or not at all.
+
+## Reptitions
+
+In both the matcher and transcriber, repetitions are indicated by placing the
+tokens to be repeated inside `$( ... )`, followed by a repetition operator,
+optionally with a separator token between. The separator token can be any token
+other than a delimiter or one of the repetition operators, but `;` and `,` are
+the most common. For instance, `$( $i:ident ),*` represents any number of
+identifiers separated by commas. Nested reptitions are permitted.
+
+The repetition operators are `*`, which indicates any number of repetitions,
+`+`, which indicates any number but at least one, and `?` which indicates an
+optional fragment with zero or one occurrences. Since `?` represents at most one
+occurrence, it cannot be used with a separator.
+
+The repeated fragment both matches and transcribes to the specified number of
+the fragment, separated by the separator token. Metavariables are matched to
+every repetition of their corresponding fragment, so for instance, the `$(
+$i:ident ),*` example above matches `$i` to all of the identifiers in the list.
+
+During transcription, additional restrictions apply to reptitions so that the
+compiler knows how to expand them properly. First, a metavariable must appear in
+exactly the same number, kind, and nesting order of repetitions in the
+transcriber as it did in the matcher. So for the matcher `$( $i:ident ),*`, the
+transcribers `=> $i`, `=> $( $( $i)* )*`, and `=> $( $i )+` are all illegal, but
+`=> $( $i );*` is correct and replaces a comma-separated list of identifiers
+with a semicolon-separated list.
+
+Second, each repetition in the transcriber must contain at least one
+metavariable in order to decide now many times to expand it. If multiple
+metavariables appear in the same repetition, they must be bound to the same
+number of fragments. For instance, `( $( $i:ident ),* ; $( $j:ident ),* ) => (
+$( ($i,$j) ),*` must bind the same number of `$i` fragments as `$j` fragments.
+This means that invoking the macro with `(a, b, c; d, e, f`) is legal and
+expands to `((a,d), (b,e), c,f))`, but `(a, b, c; d, e)` is illegal because it
+does not have the same number. This requirement applies to every layer of nested
+repetitions.
+
+> **Edition Differences**: The `?` repetition operator did not exist before the
+> 2018 edition. Prior to the 2018 Edition, `?` was an allowed
+> separator token, rather than a repetition operator.
+
+## Scoping, Exporting, and Importing
+
+For historical reasons, the scoping of macros by example does not work entirely like
```

Comment:
Or rather works almost entirely not like items.

Reply:
Almost, but not quite, entirely unlike... ;)

They do work sort of like items for path-based scoping, though, and I'd like to see us move in that direction. So I won't include this. ;)